### PR TITLE
VMware: Gather facts about vlan id from DVS Portgroup in vmware_dvs_portgroup_facts

### PR DESCRIPTION
**Pull request by dx0xm**
New arguments

##### SUMMARY
two items that I found very useful and needed to add to my project:
  

-  wouldn't be great to add a variable for particular dvswitch?

   

-  wouldn't be great to get also the vlan information of the port group?

I added it to the code

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME
vmware_dvs_portgroup_facts

### CODE REVIEW

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py]:
```suggestion
                    "vlan_id": 0
```

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py]: How about this - if user has dvSwitch with name `all`

```suggestion
        if self.dvs_name:
            # User specified specific dvswitch name to gather information
            dvsn = find_dvs_by_name(self.content, self.dvs_name)
            if dvsn is None:
                self.module.fail_json(msg="Failed to find the dvswitch %s" % self.dvs_name)
            
            dvs_lists = [dvsn]
        else:
             # default behaviour, gather information about all dvswitches
            dvs_lists = get_all_objs(self.content, [vim.DistributedVirtualSwitch], folder=datacenter.networkFolder)
```

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py]: I get 

```
AttributeError: 'vim.dvs.VmwareDistributedVirtualSwitch.PvlanSpec' object has no attribute 'vlanId'
```

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py]:
```suggestion
        dvswitch=dict(),
```

**Review by Akasurde — CHANGES_REQUESTED** [review]
Thanks for the contribution. Could you please address review comments and add test case for this change [here](https://github.com/ansible/ansible/blob/devel/test/integration/targets/vmware_dvs_portgroup_facts/tasks/main.yml) ?

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
Comment (dx0xm) [reply lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py]: hmm, how did you manage to get the error? I'm using version 6.5 and portgroups with vlan "none" give me vlan id 0.

Any suggestions, very much appreciated.

Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py]: I am using Private VLAN configuration while creating the dvpg. 

Screenshot 2019-07-12 at 11 31 26 AM


So you might want to add 
```
if isinstance(vlanInfo, vim.dvs.VmwareDistributedVirtualSwitch.PvlanSpec)
```

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
Comment (goneri) [inline lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py:229]: could you please isolate in a new function the code that you use to prepare `vlan_info`.

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
Comment (goneri) [inline lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py:255]: could you explicitly define the format of the `dvswitch` parameter: `dvswitch=dict(type='str', required=False)`

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
Comment (goneri) [inline lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py:151]: The logic used to prepare `dvs_lists` increase the global complexity of the `gather_dvs_portgroup_facts()` method. I've the feeling it would be cleaner to create a new new that init `dvs_lists` variable, and call it in the class constructor.

**Review by goneri — CHANGES_REQUESTED** [review]
Thank you @dx0xm for the patch and welcome in the Ansible contributor community :-). It needs some minor improvements before getting merged. In addition, you will need to extend the functional tests: https://github.com/ansible/ansible/blob/devel/test/integration/targets/vmware_dvs_portgroup_facts/tasks/main.yml

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py:255]: @goneri I think these are defaults so not required.

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
Comment (dx0xm) [reply lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py]: done, thanks for the hint. 
Added a new bool in the return dict:  pvlan.

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
Comment (dx0xm) [reply lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py:229]: Done.

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
Comment (dx0xm) [reply lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py:151]: what about doing all the logic in the init as implemented in vmware_local_role_facts ?

Already in the code, have a look and let me know

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
Comment (dx0xm) [reply lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py:255]: Waiting for final answer.

**Comment by dx0xm** [comment]
Something went wrong with this pull, closing it and will re open a new one shortly

**Pull request closed without merging**
